### DISCUSSION TITLE
Third-party library with no macOS platform

### DISCUSSION THREAD

**Discussion by johnrbent** [skiptools/skip#212] [Q&A] [answered]
Hello,

I am attempting to create a dual-platform wrapper library around a third-party SDK ([Mapbox Maps](https://docs.mapbox.com/)) that supplies iOS and Android variants, but is not supported on macOS. Is there a way to do this?

 I'm still learning the in's and out's of SKIP, so apologies if this is a noob question. But from what I understand Skip needs the macOS platform to transpile to Kotlin, so I can't just remove that platform. If I simply add the Mapbox dependency to Package.swift (`.product(name: "MapboxMaps", package: "mapbox-maps-ios")`) I get `While building for macOS, no library for this platform was found...` errors:

<img width="635" alt="image" src="https://github.com/user-attachments/assets/ef61477d-837b-4f66-a13e-6862294f6781">

I'm assuming this is due to Mapbox not supporting macOS. And if I edit the dependency in Package.swift to:
`.product(name: "MapboxMaps", package: "mapbox-maps-ios", condition: .when(platforms: [.iOS]))`
the import fails to find the module when building for My Mac.

I'm not sure how to proceed at the moment or if this is just a limitation of Skip. I'm hoping somebody may have come across this before and figured out a workaround.

It would be great to have a dual-platform Skip library for maps. Mapbox is an obvious choice for this.

Any advice is appreciated.

John

**Comment by marcprux** [comment]
> I'm assuming this is due to Mapbox not supporting macOS. And if I edit the dependency in Package.swift to:
> `.product(name: "MapboxMaps", package: "mapbox-maps-ios", condition: .when(platforms: [.iOS]))`
> the import fails to find the module when building for My Mac.

I think this is the way to go. Can you just guard the import in a check that the platform is not macOS? e.g.:

```
#if !os(macOS)
import MapboxMaps

… code that uses the Mapbox framework …
#endif
```

You could simply put `#if !os(macOS)` checks around each of your source files. The only reason Skip needs macOS is that there is no way to run the transpiler locally without building for the host platform. That framework doesn't actually need to *work* on macOS, just compile. We're thinking of ways that we might ease this process, since it does cause hiccups in these sorts of scenarios.

And this sounds like it will be a very useful framework! Please do let us know if we can help in any way with its development, as it will be a great addition to the Skip ecosystem.

**Reply by johnrbent** [reply]
I believe the `#if !os(macOS)` checks are removing the contained source from the transpiler - I get nearly empty transpiled files from SkipLink containing only:
```
package skip.mapbox

import skip.lib.*
```
I don't know exactly how the transpiler works, but that kind of makes sense to me that it wouldn't see inside the `#if !os(macOS)` directive when building for macOS.

I'm definitely going to dig deeper, but let me know if this is unexpected or if you have other ideas of how to work around it.

**Reply by aabewhite** [reply, marked as answer]
The transpiler's analysis of compiler directives is unfortunately pretty dumb. Basically it treats SKIP as true, and pretty much everything else as false. So I use a pattern like:

#if os(macOS) // Skip sees this as false
#else // so Skip sees this as true
...
#endif

Or combinations like:

#if SKIP || !os(macOS)
...
#endif

**Comment by johnrbent** [comment]
`#if SKIP || !os(macOS)` worked for me. Thanks for the help!
Got an nice little PoC working :) Excited to start building out the rest of the API.

<img width="400" alt="image" src="https://github.com/user-attachments/assets/e0b4d306-1023-4034-9956-facdfae1419d">

**Reply by aabewhite** [reply]
Looks great! We have a show and tell room on our Slack if you ever want to show off! https://skip.tools/slack/

**Reply by ghost** [reply]
Hi, how did you get this to work? 

Just set up an empty Xcode/Skip Project, but I am struggling with the following errors. Any help very much appreciated :))


![image](https://github.com/user-attachments/assets/9fd62ee3-494e-4fb2-b8e0-6e6f6f77fe22)


```
import SwiftUI

#if SKIP || !os(iOS)
// Skip Android Import
#else
import MapboxMaps
#endif

public struct ContentView: View {


    public init() {
    }

    public var body: some View {
        #if SKIP || !os(iOS)
            Text("Android Map")
        #else
          Map()
        #endif
        
    }
}
```

**Reply by aabewhite** [reply]
Can you also post your Package.swift? My guess would be that the problem is there. Those errors appear to be from the Swift compiler, and Skip doesn't interfere in your Swift build

**Reply by johnrbent** [reply]
Maybe try changing `#if SKIP || !os(iOS)` to `#if SKIP || !os(macOS)`?
I can't fully recall the build errors I had before setting the directive. ~~And you should only need to use that combo around the import. Your body can just use `#if SKIP ... #else ...`~~ <- That was wrong

FYI I have a skip-mapbox package in the works if you are interested in helping out with that: https://github.com/johnrbent/skip-mapbox
The APIs for the iOS and Android Mapbox SDKs are not the easiest to unify, but I've got a decent start there. Hoping to find some time soon to build it out more.

**Reply by ghost** [reply]
@aabewhite True. The mistake was laying in the Package.swift. Thanks a lot! @johnrbent Cool! I will look into it.

**Reply by ghost** [reply]
@johnrbent I looked in to it and I am wondering, how to apply your package.  

```
import SwiftUI


#if SKIP || !os(iOS)
    import SkipMapbox
#else

#endif


public struct ContentView: View {

    public init() {
    }

    public var body: some View {
        #if SKIP || !os(iOS)
            Text("Android Map")
            Map()
        #else
            Text("IOS Map")
        #endif
        
    }
}
```
I get: Unresolved reference 'mapbox' and Cannot access 'class Map<T, Output, Failure> : Publisher<T, Failure>': it is private in file. Could you maybe share an quick example with me?

**Reply by marcprux** [reply]
I think that you might be seeing a name clash here: we do have an internal [Map](https://github.com/skiptools/skip-model/blob/e15ff7e34d1bec139f42a9b303568669ddf68f09/Sources/SkipModel/Skip/Publisher.kt#L216) class in our SkipModel package which I think you may be picking up on.

If you try to fully-qualify the name `SkipMapbox.Map()`, does that change anything?

**Reply by johnrbent** [reply]
I think @marcprux is correct about the name clash and correct workaround (I noticed that during development and haven't come up with a great solution). But also I don't think I implemented that default initializer; you'll have to use the `Map(viewport)` version for now, which shouldn't clash.
I just pushed up a demo app that I've been using to test, showing the usage patterns, specifically places that need SKIP blocks: https://github.com/johnrbent/SkipMapboxDemo
See `MapDemoView`

**Reply by johnrbent** [reply]
I'm in the middle of getting the `PointAnnotation` working, but everything else should be good to go

**Reply by ghost** [reply]
@marcprux Thanks a lot! That solved it for me. @aabewhite Awesome!
